# GCC118 - Programação Matemática
## Universidade Federal de Lavras
### Instituto de Ciências Exatas e Tecnológicas
#### Profa. Andreza C. Beezão Moreira (DMM/UFLA)
#### Prof. Mayron César O. Moreira (DCC/UFLA)
#### Aluno: Bruno Crespo Ferreira

## Problema

João, Ana e Lydia são os únicos sócios e trabalhadores em uma empresa que produz relógios de primeira qualidade. João e Ana podem dedicar no máximo 40 horas por semana (cada um) à empresa, ao passo que Lydia tem disponibilidade de, no máximo, 20 horas semanais. A empresa fabrica dois tipos de relógios: relógio de pedestal (modelo antigo) e relógio de parede. Para fazer um relógio, João (que é engenheiro) monta as peças mecânicas internas do relógio, enquanto Ana (carpinteira) produz as caixas de madeira esculpidas à mão. Lydia é responsável pelas encomendas e respectiva remessa dos relógios. A quantidade de tempo necessária para cada uma dessas atividades é mostrada a seguir.

| Tarefa                           | Relógio de pedestal  | Relógio de parede |
|----------------------------------|----------------------|-------------------|
| Montagem do mecanismo do relógio | 6h                   | 4h                |
| Caixa de madeira esculpida       | 8h                   | 4h                |
| Remessa                          | 3h                   | 3h                |

Cada relógio de pedestal construído e despachado gera lucro de R\$ 300,00, ao passo que cada relógio de parede gera lucro de R\$ 200,00. Os três sócios agora querem determinar quantos relógios de cada tipo devem ser produzidos semanalmente para maximizar o lucro total.


## 1) Modelagem Matemática e Solução Ótima

### Parâmetros

* Disponibilidade de horas por semana: (40, 40, 20) para João, Ana e Lydia, respectivamente;
* A produção/entrega de um relógio de pedestal é: (6, 8, 3) horas para João, Ana e Lydia, respectivamente;
* A produção/entrega de um relógio de parede é: (4, 4, 3) horas para João, Ana e Lydia, respectivamente;
* Um relógio de pedestal gera lucro de R\$ 300,00 e um relógio de parede R\$ 200,00.


### Variáveis

* x: quantidade de relógios de pedestal produzidos semanalmente;
* y: quantidade de relógios de parede produzidos semanalmente.

### Função Objetivo

Consiste em **maximizar** o lucro total de relógios produzidos semanalmente.
\begin{equation}
max f(x, y) = 300x + 200y
\end{equation}

### Restrições

* Restrição 1: A produção dos relógios não pode ultrapassar a disponibilidade semanal de cada sócio, ou seja, teremos 3 restrições.
\begin{alignat}{2}
6x + 4y \le 40 && \quad (\text{João}) \\
8x + 4y \le 40 && \quad (\text{Ana}) \\
3x + 3y \le 20 && \quad (\text{Lydia})
\end{alignat}


### Modelo

\begin{equation}
max f(x, y) = 300x + 200y
\end{equation}

sujeito a:

\begin{alignat}{2}
6x + 4y \le 40 && \qquad \\
8x + 4y \le 40 && \qquad \\
3x + 3y \le 20 && \qquad \\
x \ge, y \ge 0 && \qquad
\end{alignat}

### Instalação da biblioteca Gurobi

In [ ]:
!pip install gurobipy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.4/14.4 MB 25.9 MB/s eta 0:00:00


### Declaração do objeto que representa o modelo matemático

In [ ]:
from gurobipy import Model, GRB

modelo = Model("relogios")

### Variáveis de decisão

In [ ]:
relogio_pedestal_var = modelo.addVar(name="relogio_pedestal", vtype=GRB.CONTINUOUS, lb=0)
relogio_parede_var = modelo.addVar(name="relogio_parede", vtype=GRB.CONTINUOUS, lb=0)

### Função objetivo

In [ ]:
modelo.setObjective(300 * relogio_pedestal_var + 200 * relogio_parede_var, GRB.MAXIMIZE)

### Restrições

In [ ]:
modelo.addConstr(6 * relogio_pedestal_var + 4 * relogio_parede_var <= 40, "restricao_joao")
modelo.addConstr(8 * relogio_pedestal_var + 4 * relogio_parede_var <= 40, "restricao_ana")
modelo.addConstr(3 * relogio_pedestal_var + 3 * relogio_parede_var <= 20, "restricao_lydia")

<gurobi.Constr *Awaiting Model Update*>

### Resolvendo o problema

In [ ]:
modelo.optimize()

Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (linux64 - "Ubuntu 22.04.3 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 3 rows, 2 columns and 6 nonzeros
Model fingerprint: 0x6103ff60
Coefficient statistics:
  Matrix range     [3e+00, 8e+00]
  Objective range  [2e+02, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+01, 4e+01]
Presolve time: 0.01s
Presolved: 3 rows, 2 columns, 6 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    5.0000000e+32   4.250000e+30   5.000000e+02      0s
       2    1.6666667e+03   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.02 seconds (0.00 work units)
Optimal objective  1.666666667e+03


### Imprimindo as soluções do problema

In [ ]:
if(modelo.status == GRB.OPTIMAL):
  print(f"Relógios de pedestal: {relogio_pedestal_var.X}")
  print(f"Relógios de parede: {relogio_parede_var.X}")
  print(f"Valor ótimo: {modelo.objVal}")

Relógios de pedestal: 3.333333333333333
Relógios de parede: 3.3333333333333335
Valor ótimo: 1666.6666666666665


## 2) Análise de Sensibilidade

In [ ]:
print("\nIntervalos de Custos (Vetor de Custos):")
for variavel in modelo.getVars():
  print(f"Variável {variavel.VarName}:")
  print(f"  lower bound do custo (SAObjLow): {variavel.SAObjLow}")
  print(f"  upper bound do custo (SAObjUp): {variavel.SAObjUp}")

print("\nIntervalos de Recursos (Vetor de Recursos):")
for restricao in modelo.getConstrs():
  print(f"Restrição {restricao.ConstrName}:")
  print(f"  lower bound do recurso (SARHSLow): {restricao.SARHSLow}")
  print(f"  upper bound do recurso (SARHSUp): {restricao.SARHSUp}")


Intervalos de Custos (Vetor de Custos):
Variável relogio_pedestal:
  lower bound do custo (SAObjLow): 200.0
  upper bound do custo (SAObjUp): 400.0
Variável relogio_parede:
  lower bound do custo (SAObjLow): 150.0
  upper bound do custo (SAObjUp): 300.0

Intervalos de Recursos (Vetor de Recursos):
Restrição restricao_joao:
  lower bound do recurso (SARHSLow): 33.333333333333336
  upper bound do recurso (SARHSUp): inf
Restrição restricao_ana:
  lower bound do recurso (SARHSLow): 26.666666666666668
  upper bound do recurso (SARHSUp): 53.33333333333333
Restrição restricao_lydia:
  lower bound do recurso (SARHSLow): 15.0
  upper bound do recurso (SARHSUp): 30.0


### Vetor de Custos

Em relação ao relógio de pedestal:
* Se seu lucro for $\lt 200$, então ele sairá da base, ou seja, somente serão produzidos os relógios de parede;
* Se for $\gt 400$, então somente ele será produzido, ou seja, não serão produzidos relógios de parede.

Em relação ao relógio de parede, idem lógica:
* Se seu lucro for $\lt 150$, então sairá da base;
* Se for $\gt 350$, então somente ele será produzido.


### Vetor de Recursos

Em relação ao João:
* Se a sua disponiblidade de trabalho alterar para menos que 33,3 horas, então algum elemento da solução básica se tornará negativo, implicando uma não factibilidade ao problema;
* E não importa o quanto trabalhe a mais, a solução ótima sempre será a mesma.

Em relação à Ana, idem lógica:
* Se sua disponibilidade de trabalho alterar para menos que 26,6 horas ou mais que 53,3 horas, então a solução ótima original não será mais válida (algum elemento se torna negativo) e o PL deve ser resolvido novamente.

Em relação à Lydia, idem lógica.

## 3) Qual sócio irá aumentar seu número máximo de horas disponíveis por semana?

A escolha se baseia naquele(a) sócio(a) que aumentaria mais o lucro total.

Tomando como base o resultado da análise de sensibilidade, sabe-se que ao João trabalhar mais não muda a solução ótima, logo ela não é um ponto da restrição de João, então escolheremos Ana ou Lydia (retas onde o ponto da solução ótima intercepta).

Ana trabalha 40 horas e seu limite superior é 53,3 horas, ou seja, um difereça de 13,3 horas. Enquanto Lydia trabalha 20 horas e seu limite superior é 30 horas, logo uma diferença de 10 horas. Considerando que o aumento ligeiro seria $1/3$ da diferença, então Ana aumentaria 4,43 horas e Lydia 3,33 horas.

Perceba que se a nova disponibilidade de Ana for 44,43 horas ou de Lydia for 23,33 horas, isso na realidade indica que o valor da função objetivo será o mesmo, pelo fato de que a solução ótima pertence às restrições de Ana e Lydia, ou seja, o "aumento ligeiro" é proporcional em $x$ e $y$.

Logicamente, Lydia deveria aumentar ligeiramente sua disponibilidade de trabalho, pois é uma diferença menor, além disso (pensando grosseiramente em um cenário realístico) faz mais sentido aumentar a carga horária de uma pessoa que trabalha 20 horas do que uma de 40 horas.



## 4) Existe alguma variável dual = 0? O que isso significa?

As variáveis duais estão relacionadas às restrições do problema. Pela análise de sensibilidade e pela resposta da pergunta (3) foi indicado que a restrição de João pode tender à mais infinito, ou seja, ele não utiliza toda a sua capacidade de horas para a produção dos relógios; matematicamente, isso indica que a restrição não está sendo utilizada (não está ativa). Portanto, mesmo não calculando podemos dizer que a variável dual de João é zero, pelo fato de não afetar, superiormente, a solução ótima.